support qnn runner multi iter run

billmguo opened the pull request:
Summary: support qnn runner multi iter run

Differential Revision: D70842764

chunit-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -167,7 +167,101 @@ void ShiftPointerIoMgr::init_io() {
+  int32_t v_cache_size = (num_heads_ + 1) * context_len_ * head_dim_;
+  int32_t k_cache_out_size = num_heads_ * max_ar_len * head_dim_;
+
+  ptr->k_cache_out.clear();
```
Hi @billmguo,
Thank you for the PR. Could you help clean up the clear and reserve functions for the iOs? From our perspective, resetting the attention mask and pointer positions should be sufficient to reset iOs.
 
Smart mask should be relatively simple. Adjust attention mask should be enough.
About ShiftPointer one please refer to prepare_kv_io and prepare_prefill_io and reassign the very beginning of each pointer to corresponding TensorImpl
 
Thanks!

billmguo replied:
1.would you comment in the code which lines can be removed?
2. About ShiftPointer one please refer to prepare_kv_io and prepare_prefill_io and reassign the very beginning of each pointer to corresponding TensorImpl
can you explain more this? do we need call prepare_kv_io and prepare_prefill_io for each generate?

billmguo replied:
Update the both shiftpointer and Smartmask logic, if you think it is still can be optimized, can you try on your side and update on PR specific code thanks!

billmguo commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -168,8 +168,35 @@ void ShiftPointerIoMgr::init_io() {
+void ShiftPointerIoMgr::reset_io(
+    const std::vector<Result<MethodMeta>>& prefill_methods_meta,
+    const std::vector<Result<MethodMeta>>& kv_methods_meta) {
+  IO* ptr = static_cast<IO*>(data_ptr_.get());
```
Initially I set like above but it is not enough

billmguo replied:
probably you can try on the device, set the above the model will fail on the 2nd rounds execution

billmguo replied:
never mind, I tried, this works I will update diff for the logics

billmguo replied:
Update the diff since the smartmask and shifpointer use different data structure for prefill and kv attn so I did not unified the reset_io into one

billmguo commented:
@pytorchbot label "topic: not user facing"

haowhsu-quic reviewed (approved):
Looks good to me, thank you!

billmguo committed:
support qnn runner multi iter run (#9071)

Summary:

support qnn runner multi iter run

Reviewed By: limintang

Differential Revision: D70842764
